src/number.rs

```diff
@@ -292,90 +391,381 @@ impl<'de> Deserialize<'de> for Number {
+            type Value = ();
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a valid number field")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<(), E>
+                where E: de::Error
+            {
+                if s == SERDE_STRUCT_FIELD_NAME {
+                    Ok(())
+                } else {
+                    Err(de::Error::custom("expected field with custom name"))
+                }
+            }
+        }
+
+        deserializer.deserialize_identifier(FieldVisitor)?;
+        Ok(NumberKey)
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+pub struct NumberFromString {
+    pub value: Number,
+}
+
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> de::Deserialize<'de> for NumberFromString {
+    fn deserialize<D>(deserializer: D) -> Result<NumberFromString, D::Error>
+    where
+        D: de::Deserializer<'de>
+    {
+        struct Visitor;
+
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = NumberFromString;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string containing a number")
+            }
+
+            fn visit_string<E>(self, s: String) -> Result<NumberFromString, E>
+            where E: de::Error,
+            {
+                Ok(NumberFromString { value: Number::from_string_unchecked(s) })
+            }
+        }
+
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+fn invalid_number() -> Error {
+    Error::syntax(ErrorCode::InvalidNumber, 0, 0)
+}
+
+macro_rules! deserialize_number {
+    ($deserialize:ident => $visit:ident) => {
+        #[cfg(not(feature = "arbitrary_precision"))]
+        fn $deserialize<V>(self, visitor: V) -> Result<V::Value, Error>
+        where
+            V: Visitor<'de>,
+        {
+            self.deserialize_any(visitor)
+        }
+
+        #[cfg(feature = "arbitrary_precision")]
+        fn $deserialize<V>(self, visitor: V) -> Result<V::Value, Error>
+        where
+            V: de::Visitor<'de>,
+        {
+            visitor.$visit(self.n.parse().map_err(|_| invalid_number())?)
+        }
+    }
+}
+
 impl<'de> Deserializer<'de> for Number {
     type Error = Error;
 
+    #[cfg(not(feature = "arbitrary_precision"))]
     #[inline]
     fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
     where
         V: Visitor<'de>,
     {
         match self.n {
-            N::PosInt(i) => visitor.visit_u64(i),
+            N::PosInt(u) => visitor.visit_u64(u),
             N::NegInt(i) => visitor.visit_i64(i),
             N::Float(f) => visitor.visit_f64(f),
         }
     }
 
+    #[cfg(feature = "arbitrary_precision")]
+    #[inline]
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
+        where V: Visitor<'de>
+    {
+        visitor.visit_map(NumberDeserializer {
```

Comment:
I think we need to be somewhat more cautious here to avoid breaking existing code. For example in the untagged enum case which relies on deserialize_any:

```rust
#[derive(Deserialize, Debug)]
#[serde(untagged)]
enum E {
    N(i64),
}

fn main() {
    println!("{:?}", E::deserialize(Number::from(0)).unwrap());
}
```

One approach would be to cascade a sequence of checks: if `self.n` can parse as u64, use visit_u64; if `self.n` can parse as i64, use visit_i64; if `self.n` parsed to f64 and then to_string'd gives you the same string as `self.n`, use visit_f64; only otherwise use visit_map.

That way for the most part only code that relies on large numbers (which previously failed to deserialize into serde_json::Number) would be affected.

Reply:
Hmm, is there something special about the way untagged enums work that causes this behaviour? It doesn't seem like the most elegant solution, but I'll have a go at this fix.